nerfstudio/models/nerfacto.py

```diff
@@ -283,7 +283,16 @@ def get_outputs(self, ray_bundle: RayBundle):
         weights_list.append(weights)
         ray_samples_list.append(ray_samples)
 
-        rgb = self.renderer_rgb(rgb=field_outputs[FieldHeadNames.RGB], weights=weights)
+        background_color = (
+            self.renderer_rgb.get_background_color(
+                rgb=field_outputs[FieldHeadNames.RGB], background_color=self.config.background_color
+            )
+            if self.training
```

Comment:
Doesn't this cause the training metrics to be still computed incorrectly? Would it be possible to defer adding the background colour to the loss computation? (Outputting RGBA instead of RGB)

Reply:
Just my two cents here: 

### For computing gradients:
Here you have to blend with the randomly chosen background color, this is the core principle of this 'carving' principle.

### For computing metrics:

Comparing RGBAs rendered by the model to the input image target RGBAs directly is not a good idea for computing metrics. If the A channel in the target image pixel is low, the RGB values don't matter, and also can't be learned by the model (since it has never seen them during training, it saw target colors blended with the randomly changing background color instead). So you'll get non-justified poor metrics for pixels with low A.

What you could do to counter this is to multiply the metric afterwards with the alpha value per pixel (alpha scaled between 0.0 and 1.0). So in short:
1. Get rendered RGBA and target pixel RGBA, where the A for the rendered pixel is accumulation. No background blending.
2. Compute metric using both RGBA values. (this will yield unjust poor metrics on pixels with low alpha).
3. Multiply the metric from step (2.) with GT alpha value (scaled between 0.0 and 1.0).

But I'm not sure how compatible this is with all metrics. So I'd still just prefer choosing a fixed (i.e. black) but configurable background color for computing metrics.

Reply:
You are right. The training metrics won't be 100% correct, and the evaluation metrics will be wrong.
Should we reintroduce the `pipeline.model.use-alpha-transparency-training` flag to determine whether to blend at this point (keeping the existing behavior) or later on (alpha-transparent training)?

Reply:
Thinking about this more carefully, the metrics will be correct once the predicted alpha matches the training alpha (desired behavior for alpha transparent training). So the metrics will still increase as expected and converge toward the correct metrics. So maybe the current implementation isn't bad, especially as computing metrics with uniform color background masking is an approximation anyway?

Reply:
I don't think computing metrics on RGBA images is a viable idea. What I suggested was the following:
1. the get_model_outputs returns RGBA when background colour is random and RGB otherwise
2. if background color is random -> compute_metrics does blending with black colour before the metrics are computed, and loss computation does blending with random colours.
3.

Reply:
Isn't this what is happening [here](https://github.com/nerfstudio-project/nerfstudio/pull/2165#discussion_r1252683799) now? `get_model_outputs` already returns accumulation (alpha channel) so no need to have an if/else logic here.